### PR TITLE
[CC-24136] Filter creds file to only required fields

### PR DESCRIPTION
**Pull request by sp-gupta**
Please refer to Jira ticket for details.

### CODE REVIEW

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/filter/GcpCredsFilter.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    // Fields allowed to pass filtering
+    private static final String[] allowedFields = new String[]
+            {"type", "project_id", "private_key_id",
+                    "private_key", "client_email", "client_id"};
+
+
+    /**
+     * This accepts the credentials config specified in the connector
+     * and returns a byte array with filtered configs.
+     * <p>Here creds config itself holds the credentials JSON string.
```
Comment (b-goyal) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/filter/GcpCredsFilter.java:30]:
```suggestion
     * <p>Here creds config itself holds the credentials JSON string or path to it.
```

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/filter/GcpCredsFilter.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+     * @param credsConfig Creds config file
```
Comment (b-goyal) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/filter/GcpCredsFilter.java:31]:
```suggestion
     * @param credsConfig Creds config file path or the stringified credentials 
```

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQueryHelper.java [hunk 3]
```diff
@@ -54,8 +56,10 @@ public BigQuery connect(String projectName, String keyFilename) {
       return connect(projectName);
     }
 
-    logger.debug("Attempting to open file {} for service account json key", keyFilename);
-    try (InputStream credentialsStream = new FileInputStream(keyFilename)) {
+    String keyfileConfig = GcpCredsFilter.filterCreds(keyFilename, true);
+
+    logger.debug("Using filtered keyfile config");
+    try (InputStream credentialsStream = new ByteArrayInputStream(keyfileConfig.getBytes(StandardCharsets.UTF_8))) {
       logger.debug("Attempting to authenticate with BigQuery using provided json key");
```
Comment (b-goyal) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQueryHelper.java:63]:
```suggestion
      logger.debug("Attempting to authenticate with BigQuery using filtered json key");
```

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQueryHelper.java [hunk 4]
```diff
@@ -54,8 +56,10 @@ public BigQuery connect(String projectName, String keyFilename) {
       return connect(projectName);
     }
 
-    logger.debug("Attempting to open file {} for service account json key", keyFilename);
-    try (InputStream credentialsStream = new FileInputStream(keyFilename)) {
+    String keyfileConfig = GcpCredsFilter.filterCreds(keyFilename, true);
+
```
Comment (b-goyal) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQueryHelper.java:60]: Can we have an empty check here

File: kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/filter/GcpCredsFilterTest.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
+package com.wepay.kafka.connect.bigquery.filter;
+
+import org.apache.commons.io.FileUtils;
+import org.junit.Test;
+
+import java.io.File;
+import java.io.IOException;
+
+import static org.junit.Assert.*;
```
Comment (b-goyal) [inline kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/filter/GcpCredsFilterTest.java:9]: Please avoid import .*

File: kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/filter/GcpCredsFilterTest.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+    @Test
+    public void testFilterCredsFile() {
+
+        String keyfile = "src/test/resources/keyfile.json";
+        String expectedKeyfileConfig = "{\n  \"type\" : \"service_account\",\n  \"project_id\" : \"connect\",\n  \"private_key_id\" : \"adjshd\",\n  \"private_key\" : \"test-key\",\n  \"client_email\" : \"test-bigquery-playground\",\n  \"client_id\" : \"1027\"\n}\n".trim();
+
+        assertEquals(expectedKeyfileConfig, GcpCredsFilter.filterCreds(keyfile, true));
+    }
+}
```
Comment (b-goyal) [inline kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/filter/GcpCredsFilterTest.java:30]: Lets add test for exception case and empty expected keyfile config.

**Review by b-goyal — COMMENTED** [review]
Some minor comments, looks good otherwise.

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/filter/GcpCredsFilter.java (hunk 1 above)
Comment (sp-gupta) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/filter/GcpCredsFilter.java:30]: done

File: kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/filter/GcpCredsFilterTest.java (hunk 5 above)
Comment (sp-gupta) [reply kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/filter/GcpCredsFilterTest.java:9]: done

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/filter/GcpCredsFilter.java (hunk 2 above)
Comment (sp-gupta) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/filter/GcpCredsFilter.java:31]: done

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQueryHelper.java (hunk 3 above)
Comment (sp-gupta) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQueryHelper.java:63]: done

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQueryHelper.java (hunk 4 above)
Comment (sp-gupta) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQueryHelper.java:60]: added it in the method itself.

File: kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/filter/GcpCredsFilterTest.java (hunk 6 above)
Comment (sp-gupta) [reply kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/filter/GcpCredsFilterTest.java:30]: done

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/filter/GcpCredsFilter.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+            // After filtering, any empty keyfile node will be translated as "{}" string
+            // which is non-empty. That's why we check whether the node is itself empty.
+            if(keyfileNode.size() == 0) {
+                throw new BigQueryConnectException("The " + KEYFILE_CONFIG +" does not contain valid fields. Please recheck the " + KEYFILE_CONFIG + " config.");
+            }
+
+            return new ObjectMapper().writer()
+                    .withDefaultPrettyPrinter().writeValueAsString(keyfileNode);
+        } catch (Exception e) {
+            throw new BigQueryConnectException("Failed to access Keyfile config: ", e);
```
Comment (b-goyal) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/filter/GcpCredsFilter.java:66]:
```suggestion
            throw new BigQueryConnectException("Failed to access " + KEYFILE_CONFIG +" config: ", e);
```

**Review by b-goyal — APPROVED** [review]
LGTM, thanks @sp-gupta
